src/rust/engine/src/tasks.rs

```diff
@@ -36,83 +35,50 @@ pub struct Tasks {
- *   1. intrinsic_add()
+ * Also has a one-shot method for adding a singleton (which has no Selectors):
+ *   1. singleton_add()
  *
  * (This protocol was original defined in a Builder, but that complicated the C lifecycle.)
  */
 impl Tasks {
   pub fn new() -> Tasks {
     Tasks {
-      intrinsics: Default::default(),
       singletons: Default::default(),
       tasks: Default::default(),
       preparing: None,
     }
   }
 
   pub fn all_product_types(&self) -> HashSet<TypeConstraint> {
-    self.all_rules().iter().map(|t| t.product)
+    self.singletons.keys().chain(self.tasks.keys())
+      .cloned()
       .collect::<HashSet<_>>()
   }
 
-  pub fn is_singleton_task(&self, sought_task: &Task) -> bool {
-    self.singletons.values().any(|tasks| tasks.iter().any(|t| t == sought_task))
-  }
-
-  pub fn is_intrinsic_task(&self, sought_task: &Task) -> bool {
-    self.intrinsics.values().any(|tasks| tasks.iter().any(|t| t == sought_task))
-  }
-
-  pub fn all_rules(&self) -> Vec<&Task> {
-    self.singletons.values()
-      .chain(self.intrinsics.values())
-      .chain(self.tasks.values())
+  pub fn all_tasks(&self) -> Vec<&Task> {
+    self.tasks.values()
       .flat_map(|tasks| tasks)
       .collect()
   }
 
-  pub fn gen_tasks(&self, subject_type: &TypeId, product: &TypeConstraint) -> Option<&Vec<Task>> {
-    // Use singletons, then intrinsics, otherwise tasks.
+  pub fn gen_singleton(&self, product: &TypeConstraint) -> Option<&(Key, Value)> {
     self.singletons.get(product)
-      .or(self.intrinsics.get(&(*subject_type, *product)))
-      .or(self.tasks.get(product))
   }
 
-  pub fn intrinsic_add(
-    &mut self,
-    func: Function,
-    // TODO: The subject_type and subject_constraint are redundant here, but we don't currently
-    // have a way to lift a TypeId into a TypeConstraint for that type.
-    subject_type: TypeId,
-    subject_constraint: TypeConstraint,
-    product: TypeConstraint
-  ) {
-    self.intrinsics.entry((subject_type, product))
-      .or_insert_with(||
-        vec![
-          Task {
-            cacheable: false,
-            product: product,
-            clause: vec![Selector::select(subject_constraint)],
-            func: func,
-          }
-        ]
-      );
+  pub fn gen_tasks(&self, product: &TypeConstraint) -> Option<&Vec<Task>> {
+    self.tasks.get(product)
   }
 
-  pub fn singleton_add(&mut self, func: Function, product: TypeConstraint) {
-    self.singletons.entry(product)
-      .or_insert_with(||
-        vec![
-          Task {
-            cacheable: false,
-            product: product,
-            clause: Vec::new(),
-            func: func,
-          }
-        ]
+  pub fn singleton_add(&mut self, value: Value, product: TypeConstraint) {
+    if let Some(&(_, ref existing_value)) = self.singletons.get(&product) {
+      panic!(
+        "More than one singleton rule was installed for the product {:?}: {:?} vs {:?}",
+        product,
+        existing_value,
+        value,
```

Comment:
I suspect that these Debug output's won't be very helpful to test against--assuming they use the derived Debug output.